man/ttulmgr.1
.TH "TTULMGR" 1 "2010-08-05" "Man Page" "Tokyo Tyrant"

.SH NAME
ttulmgr \- the command line utility of the update log API

.SH DESCRIPTION
.PP
The command `\fBttulmgr\fR' is the utility to export and import the update log.  It is useful to filter the update log with such text utilities as `\fBgrep\fR' and `\fBsed\fR'.  This command is used in the following format.  `\fIupath\fR' specifies the update log directory.
.PP
.RS
.br
\fBttulmgr export \fR[\fB\-ts \fInum\fB\fR]\fB \fR[\fB\-sid \fInum\fB\fR]\fB \fIupath\fB\fR
.RS
Export the update log as TSV text data to the standard output.
.RE
.br
\fBttulmgr import \fIupath\fB\fR
.RS
Import TSV text data from the standard input to the update log.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-ts\fR \fInum\fR : specify the beginning time stamp.
.br
\fB\-sid\fR \fInum\fR : specify the self server ID.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR ttserver (1),
.BR ttultest (1),
.BR ttutil (3),
.BR tculog (3),
.BR tcrdb (3)


ChangeLog
2010-02-02  FAL Labs  <info@fallabs.com>

	- site documents were modified.

	- Release: 1.1.41

2010-01-01  FAL Labs  <info@fallabs.com>

	* ttserver.c (proc): a bug related to initialization of the skeleton was fixed.

	* tculog.c (tculogadbaddint, tculogadbadddouble): efficiency of zero was improved.

	* tcrmgr.c (runput, procput): addint and adddouble were added.

	- Release: 1.1.40

2009-12-07  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttservstart): error handler for the accept call was added.

	* ttserver.c (proc): the system connection limit is now reset.

	- Release: 1.1.39

2009-11-05  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_mc_append, do_mc_prepend): new functions.

	- Release: 1.1.38

2009-10-20  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_stat): "fatal" attribute was added.

	* ttutil.c (ttwaitsock): new function.

	* ttutil.c (ttopensock, ttsocksend, ttsockgetc): "errno" is now copied just in case.

	* tcrdb.c (tcrdbparasearchworker): a bug of race condition was fixed.

	- Release: 1.1.37

2009-10-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_strstr): new function.

	* ttserver.c (main, proc): "-mul" option was added.

	- Release: 1.1.36

2009-09-18  FAL Labs  <info@fallabs.com>

	* tculog.c (tculogadbputshl, tculogadbputshlproc): new functions.

	* scrext.c (iterrec): a bug related to handling the return value was fixed.

	* ttserver.c (do_http_post): "tcadbmisc" is now supported.

	* ttserver.c (do_stat, do_mc_stats): counts of command execution are now output.

	* tcrdb.c (tcrdbparasearch, tcrdbparasearchworker): new functions.

	- Release: 1.1.35

2009-08-04  FAL Labs  <info@fallabs.com>

	* tcrdb.c (tcrdbput, tcrdbout, tcrdbget): reconnect mechanism was enhanced.

	* tcrdb.c (tcrdbreconnect): delay mecahism was added.

	* tculog.c (tculrdnew): a bug related to timestamp detection was fixed.

	* tcrmgr.c (runimporttsv, procimporttsv): "-sep" option was added.

	* ttserver.c (do_log): a bug of buffer overrun was fixed.

	- Release: 1.1.34

2009-07-18  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_bit): new function.

	* tcrdb.c (tcrdbqryhint, tcrdbmetasearch): new functions.

	- Release: 1.1.33

2009-07-15  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_mget, do_fwmkeys, do_misc): alignment violations were cleared.

	* tcrmttest.c (runtypical, proctypical, threadtypical): new functions.

	- Release: 1.1.32

2009-07-09  FAL Labs  <info@fallabs.com>

	* tculog.c (tculogwrite, tculrdread): the master ID is now recorded.

	* ttserver.c (main): the default value of the "-ulim" option was modified.

	* ttserver.c (proc): update logging and replicaion are now forbidden without the SID.

	* ttserver.c (do_repl): the NOP command is now throwed while circular replication.

	- Release: 1.1.31

2009-07-05  FAL Labs  <info@fallabs.com>

	* scrext.c (scrextkill): new function.

	* scrext.c (serv_lock, serv_unlock): lock mechanism was modified with TCMDB.

	- Release: 1.1.30

2009-05-28  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_http_options): new function.

	* tcrtest.c (proctable): a test of full-text search was added.

	- Release: 1.1.29

2009-05-20  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_task): a potential bug of memory leak was fixed.

	* tcrdb.c (tcrdbnew): a bug related to initialization was fixed.

	* scrext.c (serv_fwmkeys, serv_regex): new functions.

	- Release: 1.1.28

2009-05-11  FAL Labs  <info@fallabs.com>

	* tcrtest.c (runtable, proctable): "-exp" option was added.

	* ttskelnull.c: new file.

	* tokyotyrant.idl: new file.

	- Release: 1.1.27

2009-05-09  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttservsettermhandler): new function.

	* ttserver.c (do_term): new function.

	- Release: 1.1.26

2009-05-06  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttcmdidtostr): new function.

	* tculog.c (tcreplopen): a bug of data size of the server ID was fixed.

	* tcrdb.c (tcrdbsetmst, tcrdbsetmst2): the timestamp argument was added.

	* tcrmgr.c (procrepl): command names are now printed.

	* ttulmgr.c (procexport): command names are now printed.

	- Release: 1.1.25

2009-04-25  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttbreakservexpr): new function.

	* tcrdb.c (tcrdbopen2, tcrdbsetmst2, tcrdbtune): new functions.

	* tcrdb.c (tcrdbsetecode, tcrdblockmethod, tcrdbunlockmethod): new functions.

	* ttserver.c (proc): dynamic linking for the skeleton database was added.

	* ttskelmock.c, ttskelproxy.c: new files.

	- Release: 1.1.24

2009-04-07  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_eval, serv_split): new functions.

	* ttserver.c (do_repl): efficiency on dual-master replication was improved.

	* ttserver.c (do_optimize): new function.

	* ttserver.c (do_log): message format was modified.

	* tcrdb.c (tcrdboptimize): new function.

	* tcrdb.c (tcrdbvanish, tcrdbcopy, tcrdbrestore): each magic number was drifted.

	* tcrdb.c (tcrdbrestore, tcrdbsetmst): an argument for options was added.

	* tculog.c (tculogadbredo): the flag of consistency checking was separated.

	- Release: 1.1.23

2009-04-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_mapreduce): the parameter for the target keys became an option.

	- Release: 1.1.22

2009-04-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_mapreduce): the parameter for the session ID was abolished.

	- Release: 1.1.21

2009-04-01  FAL Labs  <info@fallabs.com>

	* configure.in: a bug related to support for traditional shells was fixed.

	* scrext.c (serv_stashputkeep, serv_stashputcat): new functions.

	* scrext.c (serv_mapreduce, serv_mapreducemapemit): new functions.

	* ttserver.c (do_put, do_out, do_get): parameter validation was enhanced.

	- Release: 1.1.20

2009-03-14  FAL Labs  <info@fallabs.com>

	* tcrdb.c (tcrdbqrysearchget): the option for no update log was added.

	* tcrdb.c (tcrdbqrysearchcount): new function.

	* tculog.c (tcreplread): timeout mechanism was added.

	* ttserver.c (do_repl): performance was improved.

	- Release: 1.1.19

2009-03-14  FAL Labs  <info@fallabs.com>


ttulmgr.c
/*************************************************************************************************
 * The test cases of the update log API
 *                                                               Copyright (C) 2006-2010 FAL Labs
 * This file is part of Tokyo Tyrant.
 * Tokyo Tyrant is free software; you can redistribute it and/or modify it under the terms of
 * the GNU Lesser General Public License as published by the Free Software Foundation; either
 * version 2.1 of the License or any later version.  Tokyo Tyrant is distributed in the hope
 * that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 * You should have received a copy of the GNU Lesser General Public License along with Tokyo
 * Tyrant; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA.
 *************************************************************************************************/


#include <tculog.h>
#include "myconf.h"

#define RECBUFSIZ      32                // buffer for records

typedef struct {                         // type of structure for read thread
  TCULRD *ulrd;
  int id;
  int rnum;
} TARGREAD;


/* global variables */
const char *g_progname;                  // program name


/* function prototypes */
int main(int argc, char **argv);
static void usage(void);
static void printerr(const char *msg);
static int printhex(const char *ptr, int size);
static char *mygetline(FILE *ifp);
static int runexport(int argc, char **argv);
static int runimport(int argc, char **argv);
static int procexport(const char *upath, uint64_t ts, uint32_t sid);
static int procimport(const char *upath, uint64_t lim);


/* main routine */
int main(int argc, char **argv){
  g_progname = argv[0];
  if(argc < 2) usage();
  int rv = 0;
  if(!strcmp(argv[1], "export")){
    rv = runexport(argc, argv);
  } else if(!strcmp(argv[1], "import")){
    rv = runimport(argc, argv);
  } else {
    usage();
  }
  return rv;
}


/* print the usage and exit */
static void usage(void){
  fprintf(stderr, "%s: test cases of the remote database API of Tokyo Tyrant\n", g_progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "  %s export [-ts num] [-sid num] upath\n", g_progname);
  fprintf(stderr, "  %s import upath\n", g_progname);
  fprintf(stderr, "\n");
  exit(1);
}


/* print error information */
static void printerr(const char *msg){
  fprintf(stderr, "%s: error: %s\n", g_progname, msg);
}


/* print record data */
static int printhex(const char *ptr, int size){
  int len = 0;
  while(size-- > 0){
    if(len > 0) putchar(' ');
    len += printf("%02X", *(unsigned char *)ptr);
    ptr++;
  }
  return len;
}


/* read a line from a file descriptor */
static char *mygetline(FILE *ifp){
  int len = 0;
  int blen = 1024;
  char *buf = tcmalloc(blen);
  bool end = true;
  int c;
  while((c = fgetc(ifp)) != EOF){
    end = false;
    if(c == '\0') continue;
    if(blen <= len){
      blen *= 2;
      buf = tcrealloc(buf, blen + 1);
    }
    if(c == '\n' || c == '\r') c = '\0';
    buf[len++] = c;
    if(c == '\0') break;
  }
  if(end){
    tcfree(buf);
    return NULL;
  }
  buf[len] = '\0';
  return buf;
}


/* parse arguments of export command */
static int runexport(int argc, char **argv){
  char *upath = NULL;
  uint64_t ts = 0;
  uint32_t sid = UINT32_MAX;
  for(int i = 2; i < argc; i++){
    if(!upath && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-ts")){
        if(++i >= argc) usage();
        ts = ttstrtots(argv[i]);
      } else if(!strcmp(argv[i], "-sid")){
        if(++i >= argc) usage();
        sid = tcatoi(argv[i]);
      } else {
        usage();
      }
    } else if(!upath){
      upath = argv[i];
    } else {
      usage();
    }
  }
  if(!upath) usage();
  int rv = procexport(upath, ts, sid);
  return rv;
}


/* parse arguments of import command */
static int runimport(int argc, char **argv){
  char *upath = NULL;
  uint64_t lim = 0;
  for(int i = 2; i < argc; i++){
    if(!upath && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-lim")){
        if(++i >= argc) usage();
        lim = tcatoix(argv[i]);
      } else {
        usage();
      }
    } else if(!upath){
      upath = argv[i];
    } else {
      usage();
    }
  }
  if(!upath) usage();
  int rv = procimport(upath, lim);
  return rv;
}


/* perform export command */
static int procexport(const char *upath, uint64_t ts, uint32_t sid){
  TCULOG *ulog = tculognew();
  if(!tculogopen(ulog, upath, 0)){
    printerr("tculogopen");
    return 1;
  }
  bool err = false;
  TCULRD *ulrd = tculrdnew(ulog, ts);
  if(ulrd){
    const char *rbuf;
    int rsiz;
    uint64_t rts;
    uint32_t rsid, rmid;
    while(!err && (rbuf = tculrdread(ulrd, &rsiz, &rts, &rsid, &rmid)) != NULL){
      if(rsid == sid || rmid == sid) continue;
      printf("%llu\t%u:%u\t", (unsigned long long)rts, (unsigned int)rsid, (unsigned int)rmid);
      if(rsiz >= 2){
        printf("%s\t", ttcmdidtostr(((unsigned char *)rbuf)[1]));
        printhex(rbuf, rsiz);
        putchar('\n');
      } else {
        printf("[broken entry]\n");
      }
    }
    tculrddel(ulrd);
  } else {
    printerr("tculrdnew");
    err = true;
  }
  if(!tculogclose(ulog)){
    printerr("tculogclose");
    err = true;
  }
  tculogdel(ulog);
  return err ? 1 : 0;
}


/* perform import command */
static int procimport(const char *upath, uint64_t lim){
  TCULOG *ulog = tculognew();
  if(!tculogopen(ulog, upath, lim)){
    printerr("tculogopen");
    return 1;
  }
  bool err = false;
  char *line;
  while(!err && (line = mygetline(stdin)) != NULL){
    uint64_t ts = ttstrtots(line);
    char *pv = strchr(line, '\t');
    if(!pv || ts < 1){
      tcfree(line);
      continue;
    }
    pv++;
    uint32_t sid = tcatoi(pv);
    char *mp = strchr(pv, ':');
    pv = strchr(pv, '\t');
    if(!pv){
      tcfree(line);
      continue;
    }
    pv++;
    uint32_t mid = 0;
    if(mp && mp < pv) mid = tcatoi(mp + 1);
    pv = strchr(pv, '\t');
    if(!pv){
      tcfree(line);
      continue;
    }
    pv++;
    int osiz;
    unsigned char *obj = (unsigned char *)tchexdecode(pv, &osiz);
    if(!obj || osiz < 3 || *obj != TTMAGICNUM){
      tcfree(obj);
      tcfree(line);
      continue;
    }
    if(!tculogwrite(ulog, ts, sid, mid, obj, osiz)){
      printerr("tculogwrite");
      err = true;
    }
    tcfree(obj);
    tcfree(line);
  }
  if(!tculogclose(ulog)){
    printerr("tculogclose");
    err = true;
  }
  tculogdel(ulog);
  return err ? 1 : 0;
}



// END OF FILE
